nrf_security/Kconfig

```diff
@@ -0,0 +1,477 @@
+
+choice
+	bool
+	prompt "Backend selection" if MBEDTLS_CMAC_C
+	depends on NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+	default CC310_MBEDTLS_CMAC_C
+
+	config CC310_MBEDTLS_CMAC_C
+	    bool "cc310 (CMAC-AES-128)" if (CC310_BACKEND && CC310_MBEDTLS_AES_C)
+
+	config VANILLA_MBEDTLS_CMAC_C
+	    bool "mbedtls (CMAC-AES-128, CMAC-AES-192, CMAC-AES-256)" if (VANILLA_BACKEND && VANILLA_MBEDTLS_AES_C)
+endchoice
+
+endif # AES_CONFIGURATION && NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+menuconfig AEAD_CONFIGURATION
+	bool
+	prompt "AEAD  - Authenticated Encryption with Associated Data configuration"
+	depends on CC310_BACKEND || VANILLA_BACKEND
+	default y
+
+if AEAD_CONFIGURATION
+
+if !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+config MBEDTLS_CCM_C
+	bool "AES-CCM - AES Counter with CBC-MAC mode." if MBEDTLS_AES_C
+	depends on MBEDTLS_AES_C
+	default y
+	select MBEDTLS_CIPHER_AES_256_CCM_C if VANILLA_BACKEND
+	help
+	  Enable the Counter with CBC-MAC (CCM) mode for 128-bit block cipher.
+	  This also includes CCM*
+	  MBEDTLS_CCM_C setting in mbedtls config file.
+
+endif # !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+if NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+menu "AES-CCM - Counter with CBC-MAC mode."
+	default CC310_MBEDTLS_CCM_C if (CC310_BACKEND && CC310_MBEDTLS_AES_C)
+	default VANILLA_MBEDTLS_CCM_C if (!(CC310_BACKEND && CC310_MBEDTLS_AES_C) && ((VANILLA_BACKEND && VANILLA_MBEDTLS_AES_C)))
+
+config GLUE_MBEDTLS_CIPHER_CCM_C
+	bool
+
+config CC310_MBEDTLS_CCM_C
+	bool "cc310" if (CC310_BACKEND && CC310_MBEDTLS_AES_C)
+	default y
+	help
+	  Enable the Counter with CBC-MAC (CCM) mode for 128-bit block cipher.
+	  This also includes CCM*
+	  MBEDTLS_CCM_C setting in mbedtls config file.
+
+config VANILLA_MBEDTLS_CCM_C
+	bool "mbedtls" if (VANILLA_BACKEND && VANILLA_MBEDTLS_AES_C)
+	select GLUE_MBEDTLS_CIPHER_CCM_C if CC310_MBEDTLS_CCM_C
+	select MBEDTLS_CIPHER_AES_256_CCM_C
+	help
+	  Enable the Counter with CBC-MAC (CCM) mode for 128-bit block cipher.
+	  This also includes CCM*
+	  MBEDTLS_CCM_C setting in mbedtls config file.
+endmenu
+
+endif # NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+config MBEDTLS_GCM_C
+	bool
+	prompt "AES-GCM - AES Galois/Counter Mode support"
+	default y
+	depends on VANILLA_MBEDTLS_AES_C || (VANILLA_BACKEND && MBEDTLS_AES_C)
+	help
+	  Enable the GCM module.
+	  MBEDTLS_GCM_C setting in mbedtls config file.
+
+config MBEDTLS_CHACHA20_C
+       bool
+       prompt "CHACHA-POLY mode support"
+       default y
+       depends on VANILLA_BACKEND
+       help
+	 Enable the CHACHA-POLY mode
+
+endif # AEAD_CONFIGURATION
+
+
+if !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+config MBEDTLS_DHM_C
+	bool
+	prompt "DHM   - Diffie-Hellman-Merkel modules support"
+	default y
+	depends on CC310_BACKEND || VANILLA_BACKEND
+	help
+	  Enable the DHM module.
+	  MBEDTLS_DHM_C setting in mbedtls config file.
+
+endif # !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+if NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+
+menu "DHM   - Diffie-Hellman-Merkel modules support"
```

Comment:
Should be a menuconfig?

Reply:
IMO "modules support" is superfluous.

Reply:
Will check if `menuconfig` will give the desired behavior here.

Will remove `modules support`

Reply:
The reason I'm asking :)

One backend

            *** Configuration section ***
    [*]     AES Cipher Mode configuration  --->
    [*]     AEAD  - Authenticated Encryption with Associated Data configuration  --->
    [*]     DHM   - Diffie-Hellman-Merkel modules support
    [*]     ECDH  - Elliptic Curve Diffie-Hellman library support
    [*]     ECDSA - Elliptic Curve Digital Signature Algorithm support
    [*]     RSA   - Rivest–Shamir–Adleman cryptosystem
    [*]     Hash functions configuration  --->

Both backends

            *** Configuration section ***
    [*]     AES Cipher Mode configuration  --->
    [*]     AEAD  - Authenticated Encryption with Associated Data configuration  --->
            DHM   - Diffie-Hellman-Merkel modules support  --->
            ECDH  - Elliptic Curve Diffie-Hellman library support  --->
            ECDSA - Elliptic Curve Digital Signature Algorithm support  --->
    [*]     RSA   - Rivest–Shamir–Adleman cryptosystem
                Backend selection (cc310)  --->
    [*]     Hash functions configuration  --->

Whereas I'm wondering if both could look like this:

            *** Configuration section ***
    [*]     AES   - Authenticated Encryption Standard  --->
    [*]     AEAD  - Authenticated Encryption with Associated Data  --->
    [*]     DHM   - Diffie-Hellman-Merkel  --->
    [*]     ECDH  - Elliptic Curve Diffie-Hellman  --->
    [*]     ECDSA - Elliptic Curve Digital Signature Algorithm  --->
    [*]     RSA   - Rivest–Shamir–Adleman cryptosystem  --->
    [*]     SHA   - Secure Hash Algorithm  --->

Reply:
At first, I agree with you, that last example looks nicer, but what I dislike is that it provides a false indication that, as example:
```
[*]     ECDH  - Elliptic Curve Diffie-Hellman  --->
```
give the indication that `ECDH` is enabled, which may not be the case if both backends are un-selected in the sub-menu.

Whereas the current solution does not give such indication.
Therefore it is clearer to the user that he needs to enter the sub-menu to determine if any backend is enabled.

Reply:
On the renamed menu text, I agree, text should be changed. Thanks.

Reply:
> ```
> [*]     ECDH  - Elliptic Curve Diffie-Hellman  --->
> ```
> 
> give the indication that `ECDH` is enabled, which may not be the case if both backends are un-selected in the sub-menu.
> 
> 
> Whereas the current solution does not give such indication.
> 
> Therefore it is clearer to the user that he needs to enter the sub-menu to determine if any backend is enabled.

I'd find it easier to understand if you force the selection via `depends on` (force off) or `select` (force on), so you would get 

```
        *** Configuration section ***
[*]     AES   - Authenticated Encryption Standard  --->
[*]     AEAD  - Authenticated Encryption with Associated Data  --->
[*]     DHM   - Diffie-Hellman-Merkel  --->
-*-     ECDH  - Elliptic Curve Diffie-Hellman  --->
[*]     ECDSA - Elliptic Curve Digital Signature Algorithm  --->
[*]     RSA   - Rivest–Shamir–Adleman cryptosystem  --->
[*]     SHA   - Secure Hash Algorithm  --->
```

`-*-` in menuconfig means "this has been forced on" and you can type `?` to see what is forcing it.

This patch will implement it for ECDH:

```
diff --git a/nrf_security/Kconfig b/nrf_security/Kconfig
index f9c6e69..a432992 100644
--- a/nrf_security/Kconfig
+++ b/nrf_security/Kconfig
@@ -29,6 +29,7 @@ config VANILLA_BACKEND
 # Group 0 is used to indicate when AES and ECC should use choice menu
 config NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
        bool
+       select MBEDTLS_ECDH_C
 
 # List of additional symbols to indicate cipher key sizes.
 # Currently selected based backend selection.
@@ -316,9 +317,7 @@ endmenu
 
 endif # NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
 
-if !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
-
-config MBEDTLS_ECDH_C
+menuconfig MBEDTLS_ECDH_C
        bool
        prompt "ECDH  - Elliptic Curve Diffie-Hellman library support"
        default y
@@ -327,12 +326,8 @@ config MBEDTLS_ECDH_C
          Enable the ECDH module.
          MBEDTLS_ECDH_C setting in mbedtls config file.
 
-endif # !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
-
+if MBEDTLS_ECDH_C
 if NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
-
-menu "ECDH  - Elliptic Curve Diffie-Hellman library support"
-
 config GLUE_MBEDTLS_ECDH_C
        bool
 
@@ -349,9 +344,8 @@ config VANILLA_MBEDTLS_ECDH_C
        help
          Enable the ECDH module from mbedtls vanilla.
          MBEDTLS_ECDH_C setting in mbedtls config file.
-endmenu
-
 endif # NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
+endif # MBEDTLS_ECDH_C
 
 if !NRF_CRYPTO_MULTIPLE_CHOICE_GROUP_0
 ```

Reply:
I know how kconfig works.
And what you outline here is not the desired behavior.

I think you misunderstand the purpose here.

A user can enable one or more back-end.
Depending on the back-end, different functionality might be available, such examples can:
- AES-x, e.g. AES-CBC, AES-CTR, etc. ciphers
- ECDH

as examples.

Now, AES has its own sub-menu because it contains many ciphers which can be en-/disabled individually, if the given backend supports the cipher.
Now, to make it easier for user who doesn't need anything from AES, that complete group can be disabled instead of disabling each cipher in thge sub-menu individually.

Now, ECDH, is its own, and as you noticed, if only one backend is selected, then it is a simple config, en-/disable setting.
If two (or more) backends are selected that supports ECDH, then it turns into a menu where the exact backend to use can be chosen, e.g. cc310 or mbedtls vanilla.
This is equivalent to what happens in AES under e.g. AES-CTR, AES-CBC, etc. where the same selection possibility applies if multiple backends are providing the same functionality (API wise, cause key size support may differs)

So your last example of selecting `MBEDTLS_ECDH_C` when both backends are selected is quite the opposite of what we want.